l3experimental/l3bitset/l3bitset.dtx

```diff
@@ -0,0 +1,637 @@
+%
+%    https://github.com/latex3/latex3
+%
+% for those people who are interested.
+%
+%<*driver>
+\documentclass[full]{l3doc}
+\begin{document}
+  \DocInput{\jobname.dtx}
+\end{document}
+%</driver>
+% \fi
+% \title{^^A
+%   The \pkg{l3bitset} package  \\ Bitsets ^^A
+% }
+%
+% \author{^^A
+%  The \LaTeX3 Project\thanks
+%    {^^A
+%      E-mail:
+%        \href{mailto:latex-team@latex-project.org}
+%          {latex-team@latex-project.org}^^A
+%    }^^A
+% }
+%
+% \date{Released XXXX-XX-XX}
+%
+% \maketitle
+% \begin{documentation}
+%
+% This package defines and implements the data type \texttt{bitset}, a vector of
+% bits. The size of the vector may grow dynamically.
+% Individual bits can be set and unset by index and by name.
+% The index is like all other indices in expl3 modules \emph{1-based}.
+% A \texttt{bitset} can be output as binary number or---as needed e.g. in a
+% PDF dictionary---as decimal (arabic) number.
+% Currently only a small subset of the functions provided by the \pkg{bitset}
+% package are implemented here, mainly the functions needed to use bitsets in
+% PDF dictionaries.
+%
+% The bitset is stored as a string (but one shouldn't rely on the internal
+% representation) and so the vector size is theoretically
+% unlimited, only restricted by \TeX-memory. But the export function
+% \cs{bitset_to_arabic:N} can use functions from the \texttt{int} module only if
+% the largest index used for this bitset is smaller then 32, for longer
+% bitsets \texttt{fp} is used and this is slower.
+%
+% \section{Creating bitsets}
+% \begin{function}[added = 2020-12-13]
+%   {\bitset_new:N,\bitset_new:c}
+%   \begin{syntax}
+%     \cs{bitset_new:N}  \meta{bitset var} \\
+%   \end{syntax}
+% Creates a new \meta{bitset var} or raises an error if the name is already taken.
+% The declaration is global. The \meta{bitset var} is initially 0.
+%
+%  Bitsets are implemented as string variables consisting of
+%  \texttt{1}'s and \texttt{0}'s.
+%  The rightmost number is the index position 1, so
+%  the string variable can be viewed directly as the binary number.
+%  But one shouldn't rely on the internal representation, but use the
+%  dedicated \cs{bitset_to_bin:N} instead to get the binary number.
+% \end{function}
+%
+% \begin{function}[EXP, pTF,added = 2020-12-14]
+%   { \bitset_if_exist:N, \bitset_if_exist:c  }
+%   \begin{syntax}
+%     \cs{bitset_if_exist_p:N} \meta{bitset var}
+%     \cs{bitset_if_exist:NTF} \meta{bitset var} \Arg{true code} \Arg{false code}%
+%   \end{syntax}
+%   Tests whether the \meta{bitset var} exist.
+% \end{function}
+
+%
+% \section{Setting and unsetting bits}
+%
+% \begin{function}[added = 2020-12-13]
+%   { \bitset_set_true:Nn, \bitset_set_true:cn, \bitset_gset_true:Nn, \bitset_gset_true:cn  }
+%   \begin{syntax}
+%     \cs{bitset_set_true:Nn}   \meta{bitset var}  \Arg{integer expression}\\
+%     \cs{bitset_gset_true:Nn}  \meta{bitset var}  \Arg{integer expression}
```

Comment:
Is there a practical use case for supporting "integer expressions" as the second argument or is this just the case because it happens to be technically the case?

Because if not you could think of "numerical-index" and "named-index" just being 2 ways of refering to an index position and internally handle the difference, e.g. if position is unknown, check if it is a number and if so extend otherwise complain. Then there is no need to have different commands for named or position indexes. Just thinkingout loud, it may be less efficient compared to keeping them separate, but if most of not all actual use cases use named positions because that is much more readable) it may not make any difference in efficiency.

Reply:
Well I support integer expression as all other functions which take an integer as argument support them and I didn't see a reason why not.  I can't say how many of all use cases want named positions. The examples on tex.sx I found all used simple numbers, for example https://tex.stackexchange.com/a/110886/2388.  But imho the main question about your idea to merge numbers and names is the "check if it is a number" part. How fast is it? And can it also be done for the (expandable) `\bitset_item:Nn` ?

Reply:
well that would depend on the implementation I guess. if the argument is just a key into a prop list, for example, you wouldn't need to check you would just use it (like you used the named keys, e.g. if in the named key prop you have 1=1 2=2 in addition to foo=1.  Clearly that is one step slower as it uses 2 indirection regardless of whether you use a name or a number (being now considered a name).  So I'm not saying one or the other is better, I was just offering alternatives to think about and my question was is the expression stuff ever being used because if then unifying is not an option as it means dropping the expression ability.
On the other hand, what you gain is a simpler more unified interface where names and numbers have the same status.

Reply:
just to explain my thinking a bit more (and making it easier to tell me where I went the wrong turn): assuming we have in the named prop 1=1 2=2 ... for all bitset positions that are already in use, then your current `\bitset_key:Nn` would handle both named and numbered positions. Now when you set a position to true or false you would need to check if it is a known "name" (which you currently do for named keys) and if not, only then you would need to see if your name is a number and if so  extend your bitset and add further num=num to the name prop. That would then collapes all interface commands to what you currently use with "key" in the command name and make those that currently doesn't have "key" in the command name internal ie not showing up in user/developer documentation (you would still need them to do the actual work depending on your implementation).

Reply:
Yes I understand the gain, and I like the idea in general and I'm thinking about it ;-).  But I wonder how to handle the "dynamic" growth. The bitset has no predefined length,  I don't know which 1=1, 2=2 will be needed. So one either has to drop the dynamic part and request a fix length at definition time (which imho would work too) or one has to decide what should happen if someone does `\bitset_set_true:N\l_my_bitset {42}` if the bitset itself has until now only 10 positions. How to check that 42 is an allowed position? Should one then fill the prop with 11=11, ..., 42=42?

Reply:
can't claim I have thought this through, but my initial though was that in the named prop you have 1=1 initially and whenever `\bitset_set_true:Nn` in your current implementation would require to extend the bitset I would add the corresponding num=num as well. So 
```
\bitset_new:N \l_foo_bitset                 % only 1=1
\bitset_set_trueNn\l_foo_bitset {3}    % add 2=2 and 3=3      (or maybe just 3=3? might be even better)
```

Reply:
by the way, checking for being a number really means checking for being a positive integer so that can be done very efficiently I guess.

Reply:
Well someone from the efficiency league will have to tell me a suitable test, as the only expl3 test I have is with regex: `\regex_match:nnTF { ^[\d]+$ } {#1}` (but probably it would be okay, there won't be too many tests needed).  But I think if one do it this way, then one should have a public int for a minimal length *n* (with e.g. 32 as default value), and then create the names 1=1, ... ,*n*=*n* from the start on. Should one forbid that people define names like `10=1` or must this be handled?

Reply:
Was thinking somehting low-level such as
```
\newcount\foo
\def\test#1{\afterassignment\testi\foo=0#1 \testi}
\def\testi#1\testi{\if X\detokenize{#1}X%
                   \else \foo=0 \fi
  \ifnum \foo >0  \typeout{Number}\else\typeout{no number}\fi
}
  
  \test{5}
  \test{5a}
  \test{0}
  \test{foo}
  \test{ }
  \test{1}
```
As to the other questions: I would probably not create all the immediate ones if not used and I don't think that guarding against 10=1 is needed:  (a) it would still work depending on the implementation (only be rather strange) but more importantly (b) to me this is like guarding `\int_add:Nn` having and integer expression in the second arg... if I put in "foo" there it just blows badly.

Reply:
I implemented your suggestion and imho it works fine.  Currently it uses the regex for the number test, but I will replace it by your more low-level code later.